### PR TITLE
[Java] Track pipeline options revision for idempotent initialization of file systems

### LINKED ISSUE #26321: [Bug]: FileSystems have a high risk of leaking resources due to the repeated invocation of FileSystems.setDefaultPipelineOptions
### What happened?

The underlying issue here is the repeated initialization of filesystems triggered in `SerializablePipelineOptions` (#18430).
Every call to `FileSystems.setDefaultPipelineOptions()` will silently replace any existing previous file systems without cleaning up resources.

In the case of `S3FileSystem` this can leak threads if the file system was already in use.
The simple test below shows the every increasing number of threads in the system.

```Java
public class S3FileSystemTest {
  

  @Test
  public void testSetDefaultPipelineOptions() throws Exception {
    S3Options s3Options = s3Options();
    s3Options.setAwsCredentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create("key", "secret")));

    for (int i = 0; i < 1000; i++) {
      FileSystems.setDefaultPipelineOptions(s3Options);
      FileSystems.match(ImmutableList.of("s3://bucket/key*"));
      System.out.println("Number of threads " + Thread.activeCount());
    }
  }
}  
```

This is particularly bad with some runners, e.g. see #17645:
> A normal Go Flink ValidatesRunner execution would contain the logging line from this class nearly 80,000 times, making it much more difficult to read the logs.

### Issue Priority

Priority: 3 (minor)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### LINKED ISSUE #27535: [Bug]: Repeated initialization of FileSystems is problematic
### What happened?

Current initialization of FileSystems through `FileSystems.setDefaultPipelineOptions` is problematic and prone to race conditions, especially when triggered on deserialization of `SerializablePipelineOptions` (see #18430, BEAM-14465, BEAM-14355).

Particularly with `S3FileSystem` this is bad as it can easily leak resources (threads!) if used this way (see #26321).

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by mosche**
Current initialization of FileSystems through `FileSystems.setDefaultPipelineOptions` is problematic and prone to race conditions, especially when triggered on deserialization of `SerializablePipelineOptions` (see #18430, [BEAM-14465](https://issues.apache.org/jira/browse/BEAM-14465), [BEAM-14355](https://issues.apache.org/jira/browse/BEAM-14355)).

Particularly with `S3FileSystem` this is bad as it can easily leak resources (threads!) if used this way (see #26321).

### Observations
- Removing `FileSystems.setDefaultPipelineOptions` from deserialization in SerializablePipelineOptions (#18430) is unlikely to happen any time soon as it requires a coordinated push across various runners and it’s not obvious where and when initialization is supposed to happen for each runner. As a consequence we must expect `FileSystems.setDefaultPipelineOptions` to be called any number of times.
- Looking into previous related issues ([BEAM-14465](https://issues.apache.org/jira/browse/BEAM-14465), [BEAM-14355](https://issues.apache.org/jira/browse/BEAM-14355)), logs give a clear hint how often the file systems got re-initialized. The logs mentioned in the tickets happen(ed) once per initialization of the S3 FS:  In the case of the Go Flink integration tests that was almost 80k times, running Python wordcount on Beam it was happening > 600 times.
- I spend quite some time investigating if Pipeline options could be locked during execution. That requires a lot of effort changing runners, but that part is possible. However, test cases strongly suggest being able to mutate pipeline options while running is an intended feature. 

### Changes in this PR

- Extend pipeline options to track a monotonically increasing revision with each update.
- Change the initialization of file systems to be idempotent by means of that revision. Additionally this is scoped to a particular pipeline using the options id.

Closes #27535, fixes #26321

(cc @robertwb Thanks for your reply on the email thread)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by mosche** [comment]
Run Java_IOs_Direct PreCommit

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
R: @aromanenko-dev

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java [hunk 1]
```diff
@@ -357,6 +357,8 @@ public String create(PipelineOptions options) {
    */
   Map<String, Map<String, Object>> outputRuntimeOptions();
 
+  int revision();
```
Comment (aromanenko-dev) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java:360]: Please, add a javadoc for this

**Review by aromanenko-dev — APPROVED** [review]
LGTM
Before merging, I'd ask @robertwb to take a look for additional comments

Also, it would be helpful to add some unit tests for this new "revision" functionality.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java (hunk 1 above)
Comment (mosche) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptions.java:364]: 👍 done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java [hunk 2]
```diff
@@ -103,6 +103,18 @@ public class PipelineOptionsFactoryTest {
   @Rule public TestRule restoreSystemProperties = new RestoreSystemProperties();
   @Rule public ExpectedLogs expectedLogs = ExpectedLogs.none(PipelineOptionsFactory.class);
 
+  @Test
+  public void testRevision() {
```
Comment (mosche) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:107]: @aromanenko-dev I've added a basic unit test

**Comment by aromanenko-dev** [comment]
Run Java PreCommit

**Comment by mosche** [comment]
Thanks @aromanenko-dev

**Pull request merged**